### PR TITLE
[WIP] Sink cache: fix implementation to shift key states

### LINKED ISSUE #33691: The Sink Cache looks wired
### System Info

- `transformers` version: 4.44.2
- Platform: Linux-5.15.0-86-generic-x86_64-with-glibc2.35
- Python version: 3.9.19
- Huggingface_hub version: 0.24.6
- Safetensors version: 0.4.5
- Accelerate version: 0.34.2
- Accelerate config:    not found
- PyTorch version (GPU?): 2.6.0.dev20240914 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker, @zucchini-nlp

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

https://github.com/huggingface/transformers/blob/75b7485cc72bf7122094b943af9f7d26d69ff827/src/transformers/cache_utils.py#L985-L1006

I find the sink cache does not shift the ROPE positional encoding of the new `key_states`. This would lead to
1. The positional encoding is not consecutive in all the keys;
2. In later generation rounds, the positional encoding is not like what we expected.

Since I'm not quite sure about this, I raise this issue before making a PR. Point it out if I was wrong.

### Expected behavior

It seems that after fixing this the model behavior will be more closed to [the original streamingllm implementation](https://github.com/mit-han-lab/streaming-llm/tree/main), but indeed I couldn't find a good way to test this.

I am very confused about how it could pass the test: https://github.com/huggingface/transformers/blob/317e069ee7f4d6d6595b1b03b5d9adcaede043e3/tests/utils/test_cache_utils.py#L336-L376 (though it fails on my end, the generated text is still reasonable). In this test, the `key_states` will have position encodings inrelevent to the cache size (say we have a sinkcahe with window size 256, the key state may have rope position 300. may be related #32315. If we generate texts manually so that the input positional ids and cache positions are None, the key state position will be starting from at most 257.).

### PR DESCRIPTION
**Pull request by why-in-Shanghaitech**
# What does this PR do?

Fixes #33691

This PR shifts the new key states along with those already in the cache. It could also solve some issues mentioned in #31537, #32315, where the initial input key states are longer than the context window or shorter than the attn sink.

I hope it could better explain the issue in #33691.

It does not solve the issue that the new key states have large positional encodings (See #33691 the last few lines). So it still does not work with `model.generate`. Once this is fixed, the model will be able to generate tokens very well.

Also, it will be better if we can come up with some good tests. E.g. re-design key state so that it takes the actual positional encoding as its attributes in the test?

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker @zucchini-nlp @gante
I didn't finish up reading all the previous discussions (it's too long)... But I know you should be very familiar about this.

### CODE REVIEW

**Comment by why-in-Shanghaitech** [comment]
If we make some ad-hoc changes:
https://github.com/huggingface/transformers/blob/52daf4ec768fb9ffe84a0c373834172a7c54aecc/src/transformers/models/llama/modeling_llama.py#L962
https://github.com/huggingface/transformers/blob/52daf4ec768fb9ffe84a0c373834172a7c54aecc/src/transformers/models/llama/modeling_llama.py#L967

Make this two lines always true. Then it can output the almost same texts as in [the original streamingllm implementation](https://github.com/mit-han-lab/streaming-llm/tree/main).

It might be related to the changes in #32315, but this PR will report errors from my end.

**Comment by why-in-Shanghaitech** [comment]
https://github.com/huggingface/transformers/pull/33692/commits/45c70a70c1588ae839a097088d8b36e4a688d859 This commit tries to solve the problem above. I am not sure whether this will affect other implementations, so I only modify llama as an example.

**Review by zucchini-nlp — COMMENTED** [review]
@why-in-Shanghaitech yes, SinkCache currently is very heavily bugged and the tests you mentioned are failing. We know when it started bugging, when we modified attention mask preparation stages. Prev the rerotations worked because attention mask (and consequently position ids) never went beyond the `window_length`

Currently my idea was to bring back the cropping as keeping cos/sin until the maximum possible sequence length seemed a bit inefficient in terms of memory. But I see it is also a possible solution. Since I paused work on SinkCache for a few weeks for focus on other tasks, tagging @gante 

@gante will say what is better for sink cache, he is the point of contact for anything in generation

**Comment by why-in-Shanghaitech** [comment]
> @why-in-Shanghaitech yes, SinkCache currently is very heavily bugged and the tests you mentioned are failing. We know when it started bugging, when we modified attention mask preparation stages. Prev the rerotations worked because attention mask (and consequently position ids) never went beyond the `window_length`
> 
> Currently my idea was to bring back the cropping as keeping cos/sin until the maximum possible sequence length seemed a bit inefficient in terms of memory. But I see it is also a possible solution. Since I paused work on SinkCache for a few weeks for focus on other tasks, tagging @gante
> 
> @gante will say what is better for sink cache, he is the point of contact for anything in generation

Thank you for the reply!

Yes... I think cropping is necessary since the attention mask grows w/o looking at the past key values.

https://github.com/huggingface/transformers/blob/f2c388e3f946862f657acc1e21b272ec946fc66c/src/transformers/generation/utils.py#L645-L649

Maybe we can crop here instead? The past key values are known in `model_kwargs`.

Also, since we cannot modify the positional encoding of the queries, it seems impossible to keep the cos/sin within the window length. And I think we don't need to do this? This implementation only caches cos/sin with window length + the longest prompt length (not the entire generated sequence length).

https://github.com/why-in-Shanghaitech/transformers/blob/1550f8f126946570c748131552c6b97fd33ba874/src/transformers/cache_utils.py#L969-L971

**Comment by why-in-Shanghaitech** [comment]
Is this stale?

**Comment by zucchini-nlp** [comment]
AFAIK this is not needed anymore, @gante was going to remove SinkCache so it won;t be supported anymore
